<a href="https://colab.research.google.com/github/SobokarNadiia/python_for_da_tasks/blob/main/hw_10_4_apply%2C_groupby%2C_pivot_table.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

df = pd.read_csv('/content/drive/MyDrive/dataLoves/10_pandas/data/supermarket_sales.csv')

In [2]:
df.head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3


1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [3]:
df.groupby('Branch')['gross income'].sum()

,gross income
Branch,
A,5057.1605
B,5057.0320
C,5265.1765


Найвищий прибуток має філіал С - 5265.2

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [4]:
df[df.Branch == 'C'].City.value_counts()

,count
City,
Naypyitaw,328


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [5]:
invoices_by_category = df.pivot_table(index='Branch', columns='Product line', values='Invoice ID', aggfunc='count')
invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [6]:
invoices_by_category['Electronic accessories'].sort_values(ascending=False).head(1)

,Electronic accessories
Branch,
A,60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


Три питання для аналізу:
1. Які категорії товарів приносять найбільший/найменший дохід у кожній філії?

2. Які категорії товарів мають найбільшу кількість проданих одиниць у різних філіях?

3. У яких містах і філіях середній рейтинг задоволеності покупців найвищий, а в яких — найнижчий?

1. Які категорії товарів приносять найбільший/найменший дохід у кожній філії?


In [7]:
res1 = df.pivot_table(index='Product line', columns='Branch', values='gross income', aggfunc='sum')
res1

Branch,A,B,C
Product line,,,
Electronic accessories,872.2435,811.9735,903.2845
Fashion accessories,777.7385,781.5865,1026.6700
Food and beverages,817.2905,724.5185,1131.7550
Health and beauty,599.8930,951.4600,791.2060
Home and lifestyle,1067.4855,835.6745,661.6930
Sports and travel,922.5095,951.8190,750.5680


In [8]:
def min_max_branch(val):
  return val.idxmin(), val.idxmax()

In [9]:
res1[['min', 'max']] = res1.apply(
    min_max_branch,
    axis=1,
    result_type='expand'
)

In [10]:
res1

Branch,A,B,C,min,max
Product line,,,,,
Electronic accessories,872.2435,811.9735,903.2845,B,C
Fashion accessories,777.7385,781.5865,1026.6700,A,C
Food and beverages,817.2905,724.5185,1131.7550,B,C
Health and beauty,599.8930,951.4600,791.2060,A,B
Home and lifestyle,1067.4855,835.6745,661.6930,C,A
Sports and travel,922.5095,951.8190,750.5680,C,B


In [11]:
df.groupby('Branch').City.unique()

,City
Branch,
A,[Yangon]
B,[Mandalay]
C,[Naypyitaw]


У місті Yangon (філія A) найбільший дохід приносять категорії Home and lifestyle та Sports and travel, а найменший — Health and beauty та Fashion accessories. У місті Mandalay (філія B) найбільший дохід приносять Health and beauty та Sports and travel, а найменший — Food and beverages та Electronic accessories. У місті Naypyitaw (філія C) найбільший дохід приносять Food and beverages, Fashion accessories та Electronic accessories, а найменший — Home and lifestyle та Sports and travel. На основі цих даних можна розглянути збільшення асортименту та запасів у категоріях з найвищим доходом і додатково дослідити причини низької дохідності категорій з найнижчими показниками.



---

2. Які категорії товарів мають найбільшу кількість проданих одиниць у різних філіях?

In [12]:
res2 = df.pivot_table(index='Product line', columns='Branch', values='Quantity', aggfunc='sum')
res2

Branch,A,B,C
Product line,,,
Electronic accessories,322,316,333
Fashion accessories,263,297,342
Food and beverages,313,270,369
Health and beauty,257,320,277
Home and lifestyle,371,295,245
Sports and travel,333,322,265


In [13]:
res2[['min', 'max']] = res2.apply(
    min_max_branch,
    axis=1,
    result_type='expand'
)

In [14]:
res2

Branch,A,B,C,min,max
Product line,,,,,
Electronic accessories,322,316,333,B,C
Fashion accessories,263,297,342,A,C
Food and beverages,313,270,369,B,C
Health and beauty,257,320,277,A,B
Home and lifestyle,371,295,245,C,A
Sports and travel,333,322,265,C,A


Тепер ми бачимо цікаву(або очікуєму) картину, якщо порівняти дані результати із попередніми, де ми рахували які категорії товарів приносять дохід у кожній філії, вони майже збігаються. Тобто для всіх категорій філія з найменшим валовим доходом також є філією з найменшою кількістю проданих товарів. Якщо порівнювати найбільший дохід і найбільшу продану кільксть всі катергорії окрім 'Sports and travel' також співпадають.
Ще можна порівняти із зведеною таблицею що була в завданні 3.1 де ми дивились кількість інвойсів, тобто як часто купують товри певної категорії в кожній філії.

In [15]:
invoices_by_category = invoices_by_category.T
invoices_by_category[['min', 'max']] = invoices_by_category.apply(
    min_max_branch,
    axis=1,
    result_type='expand'
)

In [16]:
invoices_by_category

Branch,A,B,C,min,max
Product line,,,,,
Electronic accessories,60,55,55,B,A
Fashion accessories,51,62,65,A,C
Food and beverages,58,50,66,B,C
Health and beauty,47,53,52,A,B
Home and lifestyle,65,50,45,C,A
Sports and travel,59,62,45,C,B


Порівняння кількості інвойсів, проданих одиниць товару та доходу показує загальну схожу тенденцію: категорії, які мають більшу кількість покупок та проданих товарів, переважно приносять і більший дохід. Водночас повного збігу між показниками немає, що може свідчити про вплив середньої вартості товарів у категорії.

In [17]:
result = pd.DataFrame({
    'Invoices': invoices_by_category['min'] + ' / ' + invoices_by_category['max'],
    'Quantity': res2['min'] + ' / ' + res2['max'],
    'Income': res1['min'] + ' / ' + res1['max']
})
result

,Invoices,Quantity,Income
Product line,,,
Electronic accessories,B / A,B / C,B / C
Fashion accessories,A / C,A / C,A / C
Food and beverages,B / C,B / C,B / C
Health and beauty,A / B,A / B,A / B
Home and lifestyle,C / A,C / A,C / A
Sports and travel,C / B,C / A,C / B


In [18]:
result['Match'] = (
    (result['Invoices'] == result['Quantity']) &
    (result['Quantity'] == result['Income'])
)

In [19]:
result

,Invoices,Quantity,Income,Match
Product line,,,,
Electronic accessories,B / A,B / C,B / C,False
Fashion accessories,A / C,A / C,A / C,True
Food and beverages,B / C,B / C,B / C,True
Health and beauty,A / B,A / B,A / B,True
Home and lifestyle,C / A,C / A,C / A,True
Sports and travel,C / B,C / A,C / B,False


У чотирьох категоріях — Fashion Accessories, Food and Beverages, Health and Beauty та Home and Lifestyle — спостерігається однакова закономірність: одні й ті самі філії мають найвищі та найнижчі значення за кількістю продажів, кількістю проданих товарів і доходом.



---

3. У яких філіях середній рейтинг задоволеності покупців найвищий, а в яких — найнижчий?

In [20]:
res3 = df.groupby('Branch').Rating.mean()
res3

,Rating
Branch,
A,7.027059
B,6.818072
C,7.072866


Рейтинг не сильно відрізняється в залежності від філії: найвищий показник має філія C — 7,07, а найнижчий — філія B — 6,82. Але всі три значення доволі низькі і варто звернути на це увагу, бо загальна задоволеність від покупки залешає бажати кращого. Треба дослідити це питання, можливо впровадити анкети для зворотнього звʼязку, щоб зрозуміти що саме не задовольняє покупців. (Це невідомо, але я припискаю що максимум шкали рейтингу - 10)



---

Отримані результати можуть бути використані для прийняття рішень щодо асортименту, управління запасами та покращення задоволеності покупців. Дані про дохід і кількість проданих товарів дозволяють визначити найбільш перспективні категорії для кожної філії та відповідно планувати запаси й асортимент. Порівняння кількості інвойсів, проданих одиниць і доходу також допомагає оцінити попит на товари та виявити категорії, де дохід може залежати не лише від обсягу продажів, а й від вартості товарів.
Аналіз рейтингу задоволеності дозволяє виявити філії, де рівень задоволеності покупців нижчий. Для таких філій доцільно провести додаткове опитування клієнтів, щоб визначити причини незадоволеності та зрозуміти, які аспекти обслуговування або асортименту потребують покращення.